### PR TITLE
posix: destroy pipe on error

### PR DESCRIPTION
**Pull request by julianuziemblo**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: `ia32-generic-qemu`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work:
    - https://github.com/phoenix-rtos/phoenix-rtos-posixsrv/pull/31 
- [ ] I will merge this PR by myself when appropriate.
@

### CODE REVIEW

File: posix/posix.c [hunk 1]
```diff
@@ -988,15 +987,15 @@ int posix_pipe(int fildes[2])
 	fo = vm_kmalloc(sizeof(open_file_t));
 	if (fo == NULL) {
 		pinfo_put(p);
-		/* FIXME: destroy pipe */
+		(void)proc_destroy(oid.port, oid);
```
Comment (ziemleszcz) [inline posix/posix.c:994]: This leaks pipes on EINTR or ENOMEM here and elsewhere. Add a FIXME/TODO comment at minimum.

Comment (julianuziemblo) [reply posix/posix.c:994]: you mean this should be done in `while (proc_destroy(oid.port, oid) == -EINTR);`? or is it a greater issue in your opinion so for now a TODO suffices?

Comment (ziemleszcz) [reply posix/posix.c:994]: `EINTR` handling is a wider kernel problem, so a TODO is acceptable. For `ENOMEM`, I’m not sure we can do much. Still, these cases should be marked with FIXME/TODO comments.

**Review by ziemleszcz — DISMISSED** [review]

**Review by oI0ck — DISMISSED** [review]
If tested then LGTM

File: posix/posix.c
Comment (ziemleszcz) [inline posix/posix.c:994]: This does not work for dummyfs. If I force an error at L617, it calls `proc_destroy()`, but the file is not removed. I suspect that unlink is required prior to destruction (unclear API). I have not checked other filesystems.

Comment (oI0ck) [reply posix/posix.c:994]: Shouldn't we move handling `proc_destroy()` to another issue?

Comment (ziemleszcz) [reply posix/posix.c:994]: The "posix/open: destroy file on error if created" commit is supposed to fix this issue. The file is not destroyed. We can remove it from this PR and address pipe/fifo issues only. Still testing it.

File: posix/posix.c [hunk 2]
```diff
@@ -664,6 +665,10 @@ int posix_open(const char *filename, int oflag, u8 *ustack)
 			return fd;
 		} while (0);
 
+		if (created != 0) {
+			(void)proc_destroy(oid.port, oid);
```
Comment (ziemleszcz) [inline posix/posix.c:671]: This does not work for dummyfs. If I force an error at L617, it calls proc_destroy(), but the file is not removed. I suspect that unlink is required prior to destruction (unclear API). I have not checked other filesystems.

File: posix/posix.c
Comment (ziemleszcz) [reply posix/posix.c:994]: Sorry, wrong location. The comment above refers to `posix_open()`.

Comment (ziemleszcz) [reply posix/posix.c:994]: I see 4 pages leak on this error path.

Comment (julianuziemblo) [reply posix/posix.c:994]: is this because of the lack of unlink, or something else?

Comment (julianuziemblo) [reply posix/posix.c:994]: about the unlink:
Unlinking it there seems to work (tested on dummyfs and ext2). I'm not sure if I should add this functionality to the proc_destroy itself (if path != NULL then proc_unlink) or should it be done separately here. My intuition is that proc_destroy seems itself as a counterpart to proc_create. 

Another issue is that mtDestroy IMO should undo everything done in mtCreate, but this is a separate implementation and interface definition issue.

For now I propose the simple pushed solution. What do you think?

Comment (ziemleszcz) [inline posix/posix.c:1014]: filedes[0] remains reserved here if `_posix_allocfd()` fails on filedes[1].

Comment (julianuziemblo) [reply posix/posix.c:1014]: hm, is it a leak though? we still have that memory and will use it when we want to open another file if I understand the code correctly?

File: posix/posix.c [hunk 3]
```diff
@@ -1059,12 +1068,15 @@ int posix_mkfifo(const char *pathname, mode_t mode)
 	/* link pipe in posix server */
 	ret = proc_link(oid, oid, pathname);
 	if (ret < 0) {
+		(void)proc_destroy(oid.port, oid);
```
Comment (ziemleszcz) [inline posix/posix.c:1071]: Same 4 pages leak on this error path as in `posix_pipe()`.

Comment (ziemleszcz) [reply posix/posix.c:1071]: FYI: I don't see this leak on next error path at L1078 so it might be connected to `unlink`.

File: posix/posix.c
Comment (ziemleszcz) [reply posix/posix.c:1014]: It is not a memory leak but the fd is reserved and cannot be assigned.

File: posix/posix.c (hunk 3 above)
Comment (oI0ck) [reply posix/posix.c:1071]: How did you manage to manage to trigger this error path, could you please send a repro?

Comment (julianuziemblo) [reply posix/posix.c:1071]: strcmp on some path in proc_link will work :D

File: posix/posix.c
Comment (ziemleszcz) [reply posix/posix.c:1014]: You can just set p->fds[filedes[]] to NULL to free it (under the lock).

Comment (ziemleszcz) [reply posix/posix.c:994]: Closing this comment in `posix_open()`. Leaving this thread open for memory leak.

Comment (ziemleszcz) [reply posix/posix.c:994]: Bug in posixsrv.

File: posix/posix.c (hunk 3 above)
Comment (ziemleszcz) [reply posix/posix.c:1071]: Bug in posixsrv.

File: posix/posix.c
Comment (julianuziemblo) [reply posix/posix.c:1014]: are you sure it remains reserved? _posix_allocfd() does not reserve anything, just returns the first empty fd (or allocates new one), and that is zeroed out (hal_memset). I don't see any more memory usage in the tests either (which would be visible if the fds are not marked as empty).

Comment (ziemleszcz) [reply posix/posix.c:1014]: Closing, descriptors are reserved in different location.

**Review by ziemleszcz — APPROVED** [review]
